Specifications/QIR/Data-Types.md

```diff
@@ -271,28 +226,28 @@ arrays:
 | __quantum__rt__array_create_1d   | `%Array* void(i32, i64)`             | Creates a new 1-dimensional array. The `i32` is the size of each element in bytes. The `i64` is the length of the array. The bytes of the new array should be set to zero. If the length is zero, the result should be an empty 1-dimensional array. |
-| __quantum__rt__array_copy        | `%Array*(%Array*)`                   | Returns a new array which is a copy of the passed-in `%Array*`. |
+| __quantum__rt__array_copy        | `%Array*(%Array*, i1)`                   | Creates a shallow copy of the array if the alias count is larger than 0 or the second argument is `true`. Returns the given array pointer otherwise, after increasing its reference count by 1. The reference count of the array elements remains unchanged. |
 | __quantum__rt__array_concatenate | `%Array*(%Array*, %Array*)`          | Returns a new array which is the concatenation of the two passed-in arrays. |
-| __quantum__rt__array_get_length  | `i64(%Array*, i32)`                  | Returns the length of a dimension of the array. The `i32` is the zero-based dimension to return the length of; it must be 0 for a 1-dimensional array. |
+| __quantum__rt__array_slice_1d       | `%Array*(%Array*, %Range, i1)`      | Creates and returns an array that is a slice of an existing 1-dimensional array. The slice may be accessing the same memory as the given array unless its alias count is larger than 0 or the last argument is `true`. The `%Range` specifies the indices that should be the elements of the returned array. The reference count of the elements remains unchanged. |
+| __quantum__rt__array_get_size_1d  | `i64(%Array*)`                  | Returns the length of a 1-dimensional array. |
 | __quantum__rt__array_get_element_ptr_1d | `i8*(%Array*, i64)`           | Returns a pointer to the element of the array at the zero-based index given by the `i64`. |
-| __quantum__rt__array_slice       | `%Array*(%Array*, i32, %Range)`      | Creates and returns an array that is a slice of an existing array. The `i32` indicates which dimension the slice is on, which must be 0 for a 1-dimensional array. The `%Range` specifies the slice. |
-| __quantum__rt__array_reference   | `void(%Array*)`                      | Indicates that a new reference has been added. |
-| __quantum__rt__array_unreference | `void(%Array*)`                      | Indicates that an existing reference has been removed and potentially releases the array. |
+| __quantum__rt__array_update_reference_count   | `void(%Array*, i32)` | Adds the given integer value to the reference count for the array. Deallocates the array if the reference count becomes 0. The behavior is undefined if the reference count becomes negative. |
+| __quantum__rt__array_update_alias_count | `void(%Array*, i32)` | Adds the given integer value to the alias count for the array. Fails if either count becomes negative. |
```

Comment:
(For negative ref count the _behavior is undefined_. But for negative alias count - _fail_)
Just out of curiosity...
My understanding is that if the count becomes negative then it means an internal bug for both cases - ref count and alias count. Then why the behavior is different (if the count becomes negative)? Why not both fail or both undefined behavior?

Reply:
I'd prefer to have both fail properly - undefined behavior is always harder to debug - but given that the change decreasing the ref count below zero may be applied to something that is already deallocated, it doesn't quite come for free to do so. Being required to properly fail I believe may have performance implications and would require additional work in the runtime; if you see any clever solutions for avoiding that, I am all ears. :)

Reply:
Thank you, Bettina. I really wasn't taking into account that the attempt to decrease the ref count down to a negative value can be accessing the _already deallocated_ instance. And such an access (even the _read_ of a ref count of the deallocated instance) already causes the _undefined behavior_ (e.g. the compiler can fully optimize out any such access if the compiler knows for sure that the instance is already deallocated).